pkg/core/state/contract_invocation.go

```diff
@@ -0,0 +1,119 @@
+package state
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/nspcc-dev/neo-go/pkg/io"
+	"github.com/nspcc-dev/neo-go/pkg/util"
+	"github.com/nspcc-dev/neo-go/pkg/vm/stackitem"
+)
+
+// NewContractInvocation return a new ContractInvocation.
+func NewContractInvocation(hash util.Uint160, method string, argBytes []byte, argCnt uint32, truncated bool) *ContractInvocation {
+	return &ContractInvocation{
+		Hash:           hash,
+		Method:         method,
+		argumentsBytes: argBytes,
+		ArgumentsCount: argCnt,
+		Truncated:      truncated,
+	}
+}
+
+// ContractInvocation contains method call information.
+// The Arguments field will be nil if serialization of the arguments exceeds the predefined limit
+// of [stackitem.MaxSerialized] (for security reasons). In that case Truncated will be set to true.
+type ContractInvocation struct {
+	Hash   util.Uint160 `json:"contract"`
+	Method string       `json:"method"`
+	// Arguments are the arguments as passed to the `args` parameter of System.Contract.Call
+	// for use in the RPC Server and RPC Client.
+	Arguments *stackitem.Array `json:"arguments"`
+	// argumentsBytes is the serialized arguments used at the interop level.
+	argumentsBytes []byte
+	ArgumentsCount uint32 `json:"argumentscount"`
+	Truncated      bool   `json:"truncated"`
+}
+
+// DecodeBinary implements the Serializable interface.
+func (ci *ContractInvocation) DecodeBinary(r *io.BinReader) {
+	ci.Hash.DecodeBinary(r)
+	ci.Method = r.ReadString()
+	if b := r.ReadVarBytes(); len(b) > 0 {
+		ci.argumentsBytes = b
+	}
```

Comment:
ci.argumentsBytes = r.ReadVarBytes()

Reply:
This is not the same behaviour. `ReadVarBytes()` will return `[]byte{}` if the length was `0`, not `nil` as expected.

`TestContractInvocation_EncodeDecode_InvalidArguments` will fail if changed to the suggestion because `WriteVarBytes` is used to the argument bytes which encodes `nil` with `0x00` but `ReadVarBytes` uses 
https://github.com/nspcc-dev/neo-go/blob/673b26fdb1097e09044776bcf46ad73627610764/pkg/io/binaryReader.go#L185

Reply:
OK, then I have even better suggestion: in `EncodeBinary` write `ci.Truncated` firstly, then write `ci.Arguments` only if `!ci.Truncated`. In `DecodeBinary` read `ci.Truncated` firstly, then use `ci.argumentsBytes = r.ReadVarBytes()` only if `!ci.Truncated`.